Refactor/23149 money request page

koko57 opened the pull request:
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. -->

### Details
<!-- Explanation of the change or anything fishy that is going on -->

### Fixed Issues
<!---
1. Please postfix `$` with a URL link to the GitHub issue this Pull Request is fixing. For example, `$ https://github.com/Expensify/App/issues/<issueID>`.
2. Please postfix  `PROPOSAL:` with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).  For example, `PROPOSAL: https://github.com/Expensify/App/issues/<issueID>#issuecomment-1369752925`

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking and its automation will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<issueID>
$ https://github.com/Expensify/App/issues/<issueID(comment)>

Do NOT only link the issue number like this: $ #<issueID>
--->
$ #23149 
PROPOSAL: [#23149(comment)](https://github.com/Expensify/App/issues/23149#issuecomment-1649617143)


### Tests
<!---
Add a numbered list of manual tests you performed that validates your changes work on all platforms, and that there are no regressions present.
Add any additional test steps if test steps are unique to a particular platform.
Manual test steps should be written so that your reviewer can repeat and verify one or more expected outcomes in the development environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Start an IOU flow (request money, split bill)
2. Verify that the value is stored properly, the whole flow works as expected and there is no other errors

- [x] Verify that no errors appear in the JS console

### Offline tests
<!---
Add any relevant steps that validate your changes work as expected in a variety of network states e.g. "offline", "spotty connection", "slow internet", etc. Manual test steps should be written so that your reviewer and QA testers can repeat and verify one or more expected outcomes. If you are unsure how the behavior should work ask for advice in the `#expensify-open-source` Slack channel.
--->

N/A

### QA Steps
<!---
Add a numbered list of manual tests that can be performed by our QA engineers on the staging environment to validate that your changes work on all platforms, and that there are no regressions present.
Add any additional QA steps if test steps are unique to a particular platform.
Manual test steps should be written so that the QA engineer can repeat and verify one or more expected outcomes in the staging environment.

For example:
1. Click on the text input to bring it into focus
2. Upload an image via copy paste
3. Verify a modal appears displaying a preview of that image
--->

1. Start an IOU flow (request money, split bill)
2. Verify that the value is stored properly, the whole flow works as expected and there is no other errors

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/28020445/a5d05d4c-e01c-4583-94c3-056a689c13e8


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/28020445/00b85b6d-197f-4a30-83fa-45d5f9ebc744


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/28020445/35ace507-1019-4ff0-90fe-8b22c28e51ae


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/28020445/079305d4-fe25-4e47-970a-5f749c7f73b6


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/28020445/25c4e80e-397e-43e9-a7ca-112dcb6abc08


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/28020445/d7dcdd62-b7b9-4768-bec0-17d87da283a4


</details>

koko57 committed:
refactor: wip - create separate new and edit pages

koko57 committed:
refactor: wip - remove refs

koko57 committed:
refactor: move util functions to a separate file, remove old MoneyRequestAmountPage

koko57 committed:
refactor: wip - remove unnecessary code in page and form components

koko57 committed:
Merge branch 'main' into refactor/23149-money-request-page

koko57 committed:
refactor: wip - proptypes for the form, cleanup

koko57 committed:
fix: fix input forwardedRef proptype

koko57 committed:
refactor: wip - proptypes, cleanup

koko57 committed:
refactor: remove old money request page

burczu commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,273 @@
+
+const amountViewID = 'amountView';
+const numPadContainerViewID = 'numPadContainerView';
+const numPadViewID = 'numPadView';
```
I would use upper case snake case for such constants

koko57 committed:
fix: fix linting error

burczu commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,273 @@
+            return;
+        }
+        event.preventDefault();
+        if (!textInput.current.isFocused()) {
```
what if `textInput.current` is `undefined`? is it possible here?

pasyukevich commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,273 @@
+     * Submit amount and navigate to a proper page
+     *
+     */
+    const handleSubmit = useCallback(() => {
```
submit handler needed to be named in another way -> https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#event-handlers

fabioh8010 commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,273 @@
+const numPadContainerViewID = 'numPadContainerView';
+const numPadViewID = 'numPadView';
+
+function MoneyRequestAmountForm(props) {
```
We should use prop destructuring -> https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#destructuring

koko57 committed:
fix: apply requested changes

fabioh8010 commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,153 @@
+    ...withCurrentUserPersonalDetailsDefaultProps,
+};
+
+function NewRequestAmountPage(props) {
```
We should use prop destructuring -> https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#destructuring

koko57 committed:
fix: apply requested changes

koko57 committed:
fix: fix forwardedRef

koko57 committed:
fix: minor fix

mountiny commented:
@koko57 Thanks for pushing this ahead. You can now merge main, the RNAnimated has been updated in main

koko57 committed:
fix: minor fix

koko57 committed:
Merge branch 'main' into refactor/23149-money-request-page

s77rt commented:
Taking this.

koko57 commented:
@s77rt @mountiny just found one bug myself, fixing it please hold on with tests until I push the fix 🙂 thanks!

koko57 committed:
fix: fix currency changing problem

koko57 committed:
fix: restore handling several tabs logic

koko57 committed:
fix: fix linting error

koko57 committed:
fix: run prettier

s77rt commented on src/pages/iou/steps/MoneyRequestAmountPage.js:
```diff
@@ -1,500 +0,0 @@
-    };
-    const titleForStep = isEditing.current ? translate('iou.amount') : title[iouType.current];
-
-    /**
-     * Check and dismiss modal
-     */
-    useEffect(() => {
-        if (!ReportUtils.shouldHideComposer(props.report, props.errors)) {
-            return;
-        }
-        Navigation.dismissModal(reportID.current);
-    }, [props.errors, props.report]);
```
Is logic no longer necessary? It was added in https://github.com/Expensify/App/pull/22198

koko57 replied:
Ok, I will restore this logic as we will be able to access error prop https://github.com/Expensify/App/pull/23394

s77rt replied:
Related PR https://github.com/Expensify/App/pull/23394. It's probably best to keep the logic

s77rt commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,192 @@
+    }, [iou.participants, iou.amount, iou.id, isEditing, iouType, reportID]);
+
+    const navigateBack = () => {
+        Navigation.goBack(isEditing.current ? ROUTES.getMoneyRequestConfirmationRoute(iouType, reportID) : null);
```
```suggestion
        Navigation.goBack(isEditing ? ROUTES.getMoneyRequestConfirmationRoute(iouType, reportID) : null);
```
No longer a ref

s77rt commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,278 @@
+    const buttonText = isEditing ? translate('common.save') : translate('common.next');
+
+    return (
+        <ScreenWrapper
+            includeSafeAreaPaddingBottom={false}
+            onEntryTransitionEnd={focusTextInput}
+        >
+            {({safeAreaPaddingBottomStyle}) => (
+                <View style={[styles.flex1, safeAreaPaddingBottomStyle]}>
+                    <HeaderWithBackButton
```
`ScreenWrapper` and `HeaderWithBackButton` should belong to `NewRequestAmountPage`

koko57 replied:
ok, so moving it out

koko57 replied:
done! 🙂

koko57 committed:
fix: rename money utils functions param, move getNewSelection to MoneyRequestAmountForm

koko57 committed:
fix: restore modal closing logic, remove current property

koko57 committed:
fix: add missing dep

koko57 committed:
refactor: move screen wrapper and header to parent component

s77rt commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -76,8 +70,6 @@ function MoneyRequestAmountForm({amount, currency, title, isEditing, onBackButto
         end: selectedAmountAsString.length,
     });
 
-    const textInput = useRef(null);
```
I don't think we should remove the internal ref and relay just on `forwardedRef`. Let's keep our ref and just forward it to `forwardedRef`.

The ref function would be similar to BaseTextInput

```js
                                ref={(ref) => {
                                    if (typeof props.innerRef === 'function') {
                                        props.innerRef(ref);
                                    } else if (props.innerRef && _.has(props.innerRef, 'current')) {
                                        // eslint-disable-next-line no-param-reassign
                                        props.innerRef.current = ref;
                                    }
                                    input.current = ref;
                                }}
```

koko57 replied:
@s77rt ok, restored!

s77rt replied:
@koko57 Please use `textInput` instead of `forwardedRef` e.g. in `onMouseDown`.

`forwardedRef` should only be referenced in the `ref` function

koko57 replied:
@s77rt fixed!

koko57 commented on src/components/AmountTextInput.js:
```diff
@@ -9,7 +9,7 @@ const propTypes = {
 
     /** A ref to forward to amount text input */
-    forwardedRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({current: PropTypes.instanceOf(React.Component)})]),
+    forwardedRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
```
But I have an error in the console. instanceOf(React.Component) doesn't seem to work with function components. And I'm passing the object ref, not a function like ref={(e) => (textInput.current = e)}

s77rt replied:
@koko57 Now we pass a function can we revert this right?

koko57 replied:
@s77rt fixed

koko57 committed:
fix: restore inner refs

koko57 committed:
fix: apply requested changes

s77rt commented:
@koko57 Please check https://github.com/Expensify/App/pull/23979#discussion_r1280895374

koko57 committed:
fix: apply requested changes

s77rt commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+    isEditing: PropTypes.bool,
+
+    /** Refs forwarded to the TextInputWithCurrencySymbol */
+    forwardedRef: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
```
```suggestion
    forwardedRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({current: PropTypes.instanceOf(React.Component)})]),
```

Do we still need to use object here? The prop type does not look correct, if there is still an error please write the ref as a function

koko57 replied:
done

koko57 committed:
fix: apply requested changes

s77rt commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+    isEditing: PropTypes.bool,
+
+    /** Refs forwarded to the TextInputWithCurrencySymbol */
+    forwardedRef: [PropTypes.func, PropTypes.shape({current: PropTypes.instanceOf(React.Component)})],
```
```suggestion
    forwardedRef: PropTypes.oneOfType([PropTypes.func, PropTypes.shape({current: PropTypes.instanceOf(React.Component)})]),
```

koko57 replied:
@s77rt sorry! 🙈 fixed

koko57 committed:
fix: minor fix

s77rt commented:
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android / native
    - [X] Android / Chrome
    - [X] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [ ] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/f772349e-435f-4f99-aa34-5c7d826e94de


</details>

<details>
<summary>Mobile Web - Chrome</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/0c053a73-7245-4675-86f7-83c2b1ae0ddb


</details>

<details>
<summary>Mobile Web - Safari</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/adb7f175-9750-4c17-a7e1-5585900e5c12


</details>

<details>
<summary>Desktop</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/c2c8afd0-afe7-4764-8c1f-b23938b93dbf


</details>

<details>
<summary>iOS</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/956bc5e2-809b-4675-9585-916743e190db


</details>

<details>
<summary>Android</summary>

<!-- add screenshots or videos here -->

https://github.com/Expensify/App/assets/16493223/5016deaa-92b7-49e5-8188-bf59ae291dce


</details>

s77rt commented:
Bug: If you change amount then change currency, the amount change is lost

1. Start money request
2. Go to edit amount flow
3. Change amount
4. Change currency

Notice that step 3 is reverted (amount is changed back)

https://github.com/Expensify/App/assets/16493223/172ed536-4149-4812-97b4-85db5813d1da

koko57 committed:
fix: bug fix

koko57 commented:
> Bug: If you change amount then change currency, the amount change is lost

@s77rt Yeah this is the bug I've fixed before but then I changed something and added a dependency unnecessarily. Fixed now

AndrewGable commented:
FYI - We've been discussing internally and we think this PR should be merged first: https://github.com/Expensify/App/pull/23046 in order to unblock development for an upcoming conference we are using this feature at. 

Sorry for the delay here, we just figured out the conflict early this morning.

koko57 commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,240 @@
+    };
+
+    return (
+        <FullPageNotFoundView shouldShow={!IOUUtils.isValidMoneyRequestType(iouType)}>
+            <ScreenWrapper
```
Are you sure? 😃 I know it makes sense, but it was implemented that way in the deleted MoneyRequestAmountPage

koko57 replied:
@s77rt I've checked how it's implemented in the other locations in the app, you're right. I applied the changes.

koko57 committed:
refactor: wrap fullpagenotfoundview into screenwrapper

marcaaron commented:
I'm going to put a HOLD on this for now. I'll jump back in here once we are ready for reviews and these more urgent changes have been incorporated. Sorry for the scope changes here @koko57. Thanks for your patience!

s77rt commented:
I think we can continue the work here @marcaaron

koko57 committed:
fix: resolve conflicts

koko57 committed:
fix: change the old money amount page for the new one

koko57 committed:
fix: remove old money amount page

koko57 committed:
fix: resolve conflicts

koko57 commented:
@s77rt @marcaaron I resolved conflicts (I replaced the MoneyRequestAmount with NewRequestAmountPage). Let me know if I need to do some more changes here 🙂

s77rt commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,242 @@
+            {({safeAreaPaddingBottomStyle}) => (
+                <FullPageNotFoundView shouldShow={!IOUUtils.isValidMoneyRequestType(iouType)}>
+                    <View style={[styles.flex1, safeAreaPaddingBottomStyle]}>
+                        {isEditing && (
+                            <HeaderWithBackButton
+                                title={titleForStep}
+                                onBackButonBackButtonPress={navigateBack}
+                            />
+                        )}
```
I didn't review in depth but I'm pretty sure we have a wrong merge conflicts here. We merged this few hours ago https://github.com/Expensify/App/pull/24097

koko57 replied:
ok, thanks @s77rt! I'll fix this

koko57 committed:
fix: render screen wrapper only when editing amount

mountiny commented:
The smartscan PR has been merged so we can proceed with this one now @s77rt @koko57 How is it looking in terms or testing and checklist? Are we close?

s77rt commented:
@mountiny I think we mostly have clean up to do. Will review asap

s77rt commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,250 @@
+                <FullPageNotFoundView shouldShow={!IOUUtils.isValidMoneyRequestType(iouType)}>
+                    <View style={[styles.flex1, safeAreaPaddingBottomStyle]}>
+                        <HeaderWithBackButton
+                            title={titleForStep}
```
I think title should always be iou.amount

koko57 replied:
Fixed!

s77rt commented:
Left one comment 😅. Will get back to this in 30min.

koko57 committed:
fix: change title for step for iou.amount

s77rt commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,242 @@
+            return;
+        }
+
+        const moneyRequestID = `${iouType}${reportID}`;
+        const shouldReset = iou.id !== moneyRequestID;
+        // If the money request ID in Onyx does not match the ID from params, we want to start a new request
+        // with the ID from params. We need to clear the participants in case the new request is initiated from FAB.
+        if (shouldReset) {
+            IOU.setMoneyRequestId(moneyRequestID);
+            IOU.setMoneyRequestDescription('');
+            IOU.setMoneyRequestParticipants([]);
+        }
+
+        // If a request is initiated on a report, skip the participants selection step and navigate to the confirmation page.
+        if (report.reportID) {
+            // Reinitialize the participants when the money request ID in Onyx does not match the ID from params
+            if (_.isEmpty(participants)) {
+                const currentUserAccountID = currentUserPersonalDetails.accountID;
+                const iouParticipants = ReportUtils.isPolicyExpenseChat(report)
+                    ? [{reportID: report.reportID, isPolicyExpenseChat: true, selected: true}]
+                    : _.chain(report.participantAccountIDs)
+                          .filter((accountID) => currentUserAccountID !== accountID)
+                          .map((accountID) => ({accountID, selected: true}))
+                          .value();
+                IOU.setMoneyRequestParticipants(iouParticipants);
+            }
+            Navigation.navigate(ROUTES.getMoneyRequestConfirmationRoute(iouType, reportID));
+            return;
+        }
+        Navigation.navigate(ROUTES.getMoneyRequestParticipantsRoute(iouType));
```
```suggestion
        IOU.navigateToNextPage(iou, iouType, reportID, report);
```

koko57 replied:
nice 😃

s77rt commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,242 @@
+        />
+    );
+
+    // ScreenWrapper is only needed in edit mode because we have a dedicated route for the edit amount page (MoneyRequestEditAmountPage).
+    // The rest of the cases this component is rendered through <MoneyRequestSelectorPage /> which has it's own ScreenWrapper
+    if (!isEditing) {
+        return content;
+    }
+
+    return (
+        <ScreenWrapper
+            includeSafeAreaPaddingBottom={false}
+            onEntryTransitionEnd={focusTextInput}
```
NAB I think we want to split this into two pages, one that uses `ScreenWrapper` for `MoneyRequestEditAmountPage` screen. And the other that does not (meant to be a part of the page) for `MoneyRequestSelectorPage`.

I'm questioning this change as the conditional rendering above just seems easier :grin:

koko57 replied:
TBH I'm not sure if I can visualise this change right now 😅 Maybe this could be done later on?

koko57 committed:
fix: apply requested changes

s77rt reviewed (approved):
LGTM! :rocket:

marcaaron commented on src/libs/MoneyRequestUtils.js:
```diff
@@ -0,0 +1,82 @@
+ * @param {String} amount
+ * @returns {String}
+ */
+const stripCommaFromAmount = (amount) => amount.replace(/,/g, '');
```
I believe we prefer `function` over arrow functions in lib files like this.

marcaaron commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+        saveAmountToState(currency, amount);
+        // we want to update the state only when the amount is changed
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [amount]);
```
**NAB/possible follow up:** 

It seems a bit strange to keep the formatted amount in state. Could the formatted amount also just be computed when the `props.amount` change? That kind of makes more sense to me since:

- `props.amount` will change
- recompute the formatted amount on render (possibly optimized with useMemo if necessary)

Something like that seems like it would avoid an additional render from setting the state immediately each time this prop changes.

Maybe the logic related to the selection stuff makes this more difficult. I would try to compute that stuff as well if possible. I'm not sure there is a clear reason why it needs to be in state.

mountiny replied:
+1 to this comment, we need to keep the amount and currency as separate properties in the transaction object so I think it would make sense to not format the amount til we need to display it

koko57 replied:
@mountiny @marcaaron because of the selection logic I think we should keep it.  we have to keep the value of the text input somehow and it’s easier to keep it that way (as a string) as changing the value will occur more often by changing the input value than by changing the prop. The selection logic is used in both TextInputWithCurrencySymbol and BigNumberPad

marcaaron commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+     * Sets the state according to amount that is passed
+     * @param {String} newAmount - Changed amount from user input
+     */
+    const setNewAmount = (newAmount) => {
```
NAB, I wonder if we can make all the methods related to amounts more descriptive. What's the difference between "setting the new amount" vs. "setting the current amount" vs. "updating the amount" vs. "saving the amount to state". I still feel like we have some really heavy logic all related to the "amount" here and that was one of the main things I would still like to see addressed as part of this code cleanup.

koko57 replied:
I'd move this addLeadingZero to TextInputWithCurrencySymbol (and updateAmount could be removed) , and remove saveAmountToState -> it’s only used in one useEffect, it can be done inside the hook

marcaaron commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+     * @param {Boolean} value - Changed text from user input
+     */
+    const updateLongPressHandlerState = useCallback((value) => {
+        setShouldUpdateSelection(!value);
```
**NAB/future consideration** but it would be great if we could somehow uncouple this selection logic. It feels like the `TextInputWithCurrencySymbol` should maybe be handling this stuff, but not sure if it's used anywhere else atm.

marcaaron commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+     *
+     * @param {String} text - Changed text from user input
+     */
+    const updateAmount = (text) => {
```
NAB, another method that will do _something_ with the amount. I feel like there are too many and we should make them more descriptive.

marcaaron commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,211 @@
+
+    const currency = currentCurrency || iou.currency;
+
+    /**
+     * Focus text input
+     */
```
```suggestion
```

marcaaron commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,211 @@
+        Navigation.dismissModal(reportID);
+    }, [errors, report, reportID]);
+
+    // Because we use Onyx to store iou info, when we try to make two different money requests from different tabs, it can result in many bugs.
```
>it can result in many bugs.

NAB, it would be better to say which ones if possible. People won't know which bugs are fixed - and won't think of any alternative solutions or be able to test if the logic works as intended or can be removed in the future, etc.

The code here is still not very intuitive or explained well.

marcaaron commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,211 @@
+    }, [errors, report, reportID]);
+
+    // Because we use Onyx to store iou info, when we try to make two different money requests from different tabs, it can result in many bugs.
+    // This logic is added to prevent such bugs.
```
```suggestion
```

marcaaron commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,211 @@
+    useEffect(() => {
+        if (isEditing) {
+            if (prevMoneyRequestID.current !== iou.id) {
+                // The ID is cleared on completing a request. In that case, we will do nothing.
+                if (iou.id) {
+                    Navigation.goBack(ROUTES.getMoneyRequestRoute(iouType, reportID), true);
+                }
+                return;
```
```suggestion
                // The ID is cleared on completing a request. In that case, we will do nothing.
                if (!iou.id) {
                    return;
                }
                
                Navigation.goBack(ROUTES.getMoneyRequestRoute(iouType, reportID), true);
                return;
```

marcaaron reviewed (changes requested):
I think this PR is a great step in the right direction so labeling most of my comments as non-blockers (I think we can possibly save them for a follow up PR). Anything not labeled with a NAB would be good to fix.

Thanks for the changes @koko57 !

koko57 committed:
fix: apply requested changes

koko57 commented:
@marcaaron thanks for the review! 😊 I fixed all the code with comments without NAB label.

koko57 commented:
@marcaaron @mountiny Can I start working on this followup PR?

marcaaron reviewed (approved):
Thanks for the changes. 👍 from me.

marcaaron commented:
@koko57 I'd say go for it unless @mountiny wants to hold off for his next round of changes (maybe this is easier for everyone).

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

mountiny commented:
@koko57 Yes please, thanks everyone for pushing this ahead, no need to wait with the follow ups

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5788511088) to staging by https://github.com/marcaaron in version: 1.3.51-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5806686683) to production by https://github.com/Julesssss in version: 1.3.51-2 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5806699766) to staging by https://github.com/marcaaron in version: 1.3.52-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|failure ❌
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5806699766) to staging by https://github.com/marcaaron in version: 1.3.52-0 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/5825482823) to production by https://github.com/puneetlath in version: 1.3.52-5 🚀

 platform | result 
 ---|--- 
🤖 android 🤖|success ✅ 
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅ 
🕸 web 🕸|success ✅

eVoloshchak commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+        }
+        setCurrentAmount((prevAmount) => {
+            setSelection((prevSelection) => getNewSelection(prevSelection, prevAmount.length, newAmountWithoutSpaces.length));
+            return MoneyRequestUtils.stripCommaFromAmount(newAmountWithoutSpaces);
```
👋 Coming from https://github.com/Expensify/App/issues/23933
> Cursor jumps to next position when we use ',' in request money when language is English

This happens because we were updating the cursor position using the non-comma stripped amount and only after that strip the comma from the amount.
We resolved this by stripping comma from the amount first and updating the selection after that, i.e
```js
  setCurrentAmount((prevAmount) => {
      const strippedAmount = MoneyRequestUtils.stripCommaFromAmount(newAmountWithoutSpaces);
      setSelection((prevSelection) => getNewSelection(prevSelection, prevAmount.length, strippedAmount.length));
      return strippedAmount;
  });

s77rt replied:
I think that makes sense but it's not a regression from this PR. That logic was copied from `src/pages/iou/steps/MoneyRequestAmount.js`,

sobitneupane commented:
[This issue](https://github.com/Expensify/App/issues/26619) was caused by change introduced by this PR.

Issue: Web - App crashes when currency in url is changed to invalid currency code 

Steps for reproduction:

1. Go to any chat
2. Open request money
3. Change currency in currency list
4. Notice there is a currency code in url indicating selected currency
5. (a). Now change that currency code in url to invalid currency code for example ABCD
     (b). Change currency code in url to not invalid currency code but keep it 3 letters like ABC

koko57 commented:
@sobitneupane I don't think this PR introduced the issue 🙂 We didn't validate the currency before these changes were introduced
![Screenshot 2023-10-09 at 10 03 17](https://github.com/Expensify/App/assets/28020445/d414fcc7-7d6c-42e2-94ca-d21d2c55fa51)

eVoloshchak commented on src/pages/iou/steps/NewRequestAmountPage.js:
```diff
@@ -0,0 +1,208 @@
+
+    useFocusEffect(
+        useCallback(() => {
+            focusTextInput();
```
Coming from https://github.com/Expensify/App/issues/22388, this has caused a small regression 
We're focusing the input immediately after the page is focused. When this happens when the page starts navigating, it will cancel the transition animation and we just see an abrupt page change

It was later decided to use [this solution](https://expensify.slack.com/archives/C01GTK53T8Q/p1694660990479979) whenever we need to autofocus a TexInput, which doesn't have this problem

koko57 replied:
@eVoloshchak It was already implemented earlier https://github.com/Expensify/App/blob/e2694f52ce61fa9e67ee7ab1636cc38b2ba24d48/src/pages/iou/steps/MoneyRequestAmount.js#L292-L296. It was just moved to the new file 🙂

eVoloshchak replied:
Oops, I see, thank you!

koko57 commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+                ) : null}
+                <Button
+                    success
+                    style={[styles.w100, styles.mt5]}
```
it wasn't added in this commit 😅 it was just moved from the old file without changes https://github.com/Expensify/App/blob/e2694f52ce61fa9e67ee7ab1636cc38b2ba24d48/src/pages/iou/steps/MoneyRequestAmount.js#L462

koko57 commented on src/pages/iou/steps/MoneyRequestAmount.js:
```diff
@@ -1,510 +0,0 @@
-    const textInput = useRef(null);
-    const iouType = useRef(lodashGet(props.route, 'params.iouType', ''));
-    const reportID = useRef(lodashGet(props.route, 'params.reportID', ''));
-    const isEditing = useRef(lodashGet(props.route, 'path', '').includes('amount'));
```
@0xmiroslav - https://github.com/Expensify/App/pull/23979#discussion_r1412506814 - it's not a regression from this PR ;) it was  introduced earlier, logic was just moved there. (the issue wasn't a problem back then)

situchan commented on src/pages/iou/steps/MoneyRequestAmountForm.js:
```diff
@@ -0,0 +1,266 @@
+    }, [onSubmitButtonPress, currentAmount]);
+
+    const formattedAmount = MoneyRequestUtils.replaceAllDigits(currentAmount, toLocaleDigit);
+    const buttonText = isEditing ? translate('common.save') : translate('common.next');
```
This component is used in `EditRequestAmountPage` as well. `isEditing` prop is not passed so it caused https://github.com/Expensify/App/issues/33011

koko57 replied:
@situchan this prop was passed properly in EditRequestAmountPage, but it was lost during refactor https://github.com/Expensify/App/pull/28618/files#diff-598c08eb72e9071a7ab827304a24f3dd0a769b65f71e9c30140df3eda8c4b667
